File: dart/bf.dart
#!/usr/bin/env dart
import 'dart:io';
import 'dart:typed_data';

void main(List<String> arguments) {
  var program;
  if (arguments.isEmpty) {
    var buffer = new StringBuffer();
    var line;
    while ((line = stdin.readLineSync()) != null) {
      buffer.write(line);
    }
    program = buffer.toString();
  } else {
    program = new File(arguments[0]).readAsStringSync();
  }

  program = program.replaceAll(RegExp(r'[^><+\-.,[\]]'), '');
  var len = program.length;
  var jumps = new List(len);
  var stack = [];

  for (var i = 0; i < len; ++i) {
    switch (program[i]) {
      case '[':
        stack.add(i);
        break;
      case ']':
        var j = stack.removeLast();
        jumps[j] = i;
        jumps[i] = j;
        break;
    }
  }

  var tape = new Uint8List(65536);
  var ptr = 0;

  for (var i = 0; i < len; ++i) {
    switch (program[i]) {
      case '>':
        ptr = ptr+1 & 65535;
        break;
      case '<':
        ptr = ptr-1 & 65535;
        break;
      case '+':
        ++tape[ptr];
        break;
      case '-':
        --tape[ptr];
        break;
      case '.':
        stdout.writeCharCode(tape[ptr]);
        break;
      case ',':
        var c = stdin.readByteSync();
        if (c != -1) {
          tape[ptr] = c;
        }
        break;
      case '[':
        if (tape[ptr] == 0) {
          i = jumps[i];
        }
        break;
      case ']':
        if (tape[ptr] != 0) {
          i = jumps[i];
        }
        break;
    }
  }
}

File: swift/bf.swift
#!/usr/bin/env swift
import Foundation

var program = ""

if CommandLine.argc > 1 {
    do {
        program = try String(contentsOfFile: CommandLine.arguments[1])
    } catch let error as NSError {
        let msg = "\(CommandLine.arguments[1]): \(error.localizedDescription)\n"

        FileHandle.standardError
                  .write(msg.data(using: .utf8)!)

        exit(1)
    }
} else {
    while let line = readLine() {
        program.append(line)
    }
}

program = program.filter("><+-.,[]".contains)

var jumps = Array(repeating: 0, count: program.count)
var stack: [Int] = []

for (i, c) in program.enumerated() {
    if c == "[" {
        stack.append(i)
    } else if c == "]" {
        if stack.count == 0 {
            FileHandle.standardError
                      .write("error: extraneous ']'\n".data(using: .utf8)!)

            exit(1)
        }

        let j = stack.removeLast()

        jumps[i] = j
        jumps[j] = i
    }
}

if stack.count > 0 {
    FileHandle.standardError
              .write("error: expected ']'\n".data(using: .utf8)!)

    exit(1)
}

let len = program.count

var i = 0
var line = ""
var ptr = 0
var tape = Array(repeating: 0, count: 65536)

while (i < len) {
    let c = program[program.index(program.startIndex, offsetBy: i)]

    if c == ">" {
        ptr = (ptr + 1) & 65535
    } else if c == "<" {
        ptr = (ptr - 1) & 65535
    } else if c == "+" {
        tape[ptr] = (tape[ptr] + 1) & 255
    } else if c == "-" {
        tape[ptr] = (tape[ptr] - 1) & 255
    } else if c == "." {
        print(Character(UnicodeScalar(tape[ptr])!), terminator: "")
    } else if c == "," {
        if line.count == 0 {
            let input = readLine(strippingNewline: false)

            if input != nil {
                line = input!
            }
        }

        if line.count > 0 {
            tape[ptr] = Int(UnicodeScalar(String(line
                            .remove(at: line.startIndex)))!.value)
        }
    } else if c == "[" {
        if tape[ptr] == 0 {
            i = jumps[i]
        }
    } else if c == "]" {
        if tape[ptr] != 0 {
            i = jumps[i]
        }
    }


    i += 1
}
